File: src/Application Specific/Classes/Entity Collections/CostCentres/CostCentreSql.vb
Imports System.Collections.Generic
Imports System.Linq
Imports FSM.Entity.CostCentres.Enums
Imports FSM.Entity.Sys

Namespace Entity.CostCentres

    Public Class CostCentreSql

        Private _database As Sys.Database

        Public Sub New(ByVal database As Sys.Database)
            _database = database
        End Sub

        Public Function [GetAll]() As DataView
            _database.ClearParameter()
            Return New DataView(_database.ExecuteSP_DataTable("CostCentre_GetAll"))
        End Function

        Public Function [Get](ByVal ref As Integer, ByVal ref2 As Integer, ByVal getBy As GetBy) As List(Of CostCentre)
            _database.ClearParameter()

            'Get the datatable from the database store in dt
            Dim dt As DataTable = Nothing
            Select Case getBy
                Case GetBy.Id
                    _database.AddParameter("@Id", ref, True)
                    dt = _database.ExecuteSP_DataTable("CostCentre_Get_ById")

                Case GetBy.JobId
                    _database.AddParameter("@JobID", ref, True)
                    dt = _database.ExecuteSP_DataTable("CostCentre_Get_ByJobId")

                Case GetBy.JobTypeId
                    _database.AddParameter("@JobTypeId", ref, True)
                    dt = _database.ExecuteSP_DataTable("CostCentre_Get_ByJobTypeId")

                Case GetBy.JobTypeIdAndCutomerId
                    _database.AddParameter("@JobTypeId", ref, True)
                    _database.AddParameter("@CustomerId", ref2, True)
                    dt = _database.ExecuteSP_DataTable("CostCentre_Get_ByJobTypeIdAndCustomerId")

                Case Else
                    _database.AddParameter("@Id", ref, True)
                    dt = _database.ExecuteSP_DataTable("CostCentre_Get_ById")
            End Select

            If dt IsNot Nothing AndAlso dt.Rows.Count > 0 Then
                Dim costCentres As List(Of CostCentre) = ObjectMap.DataTableToList(Of CostCentre)(dt)
                Return costCentres
            Else : Return Nothing
            End If
        End Function

        Public Function [Insert](ByVal ccm As CostCentre) As CostCentre
            AddParameters(ccm)
            ccm.Id = Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("CostCentre_Insert"))
            Return ccm
        End Function

        Private Sub AddParameters(ByVal ccm As CostCentre)
            _database.ClearParameter()
            _database.AddParameter("@CostCentre", ccm.CostCentre)
            _database.AddParameter("@JobTypeId", ccm.JobTypeId)
            _database.AddParameter("@LinkId", ccm.LinkId)
            _database.AddParameter("@LinkTypeId", ccm.LinkTypeId)
            Dim spendLimit As Object = If(ccm.JobSpendLimit > 0, CObj(ccm.JobSpendLimit), DBNull.Value)
            _database.AddParameter("@JobSpendLimit", spendLimit, True)
        End Sub

        Public Function [Update](ByVal ccm As CostCentre) As CostCentre
            AddParameters(ccm)
            _database.AddParameter("@Id", ccm.Id)
            ccm.Id = Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("CostCentre_Update"))
            Return ccm
        End Function

        Public Function [Delete](ByVal Id As Integer) As Boolean
            _database.ClearParameter()
            _database.AddParameter("@Id", Id)
            Return CBool(_database.ExecuteSP_ReturnRowsAffected("CostCentre_Delete") = 1)
        End Function

    End Class

End Namespace

File: src/Application Specific/Classes/Entity Collections/CostCentres/CostCentreSql.cs
using System;
using System.Collections.Generic;
using System.Data;
using FSM.Entity.CostCentres.Enums;
using FSM.Entity.Sys;
using Microsoft.VisualBasic.CompilerServices;

namespace FSM.Entity.CostCentres
{
    public class CostCentreSql
    {
        private Database _database;

        public CostCentreSql(Database database)
        {
            _database = database;
        }

        public DataView GetAll()
        {
            _database.ClearParameter();
            return new DataView(_database.ExecuteSP_DataTable("CostCentre_GetAll"));
        }

        public List<CostCentre> Get(int @ref, int ref2, GetBy getBy)
        {
            _database.ClearParameter();

            // Get the datatable from the database store in dt
            DataTable dt = null;
            switch (getBy)
            {
                case GetBy.Id:
                    {
                        _database.AddParameter("@Id", @ref, true);
                        dt = _database.ExecuteSP_DataTable("CostCentre_Get_ById");
                        break;
                    }

                case GetBy.JobId:
                    {
                        _database.AddParameter("@JobID", @ref, true);
                        dt = _database.ExecuteSP_DataTable("CostCentre_Get_ByJobId");
                        break;
                    }

                case GetBy.JobTypeId:
                    {
                        _database.AddParameter("@JobTypeId", @ref, true);
                        dt = _database.ExecuteSP_DataTable("CostCentre_Get_ByJobTypeId");
                        break;
                    }

                case GetBy.JobTypeIdAndCutomerId:
                    {
                        _database.AddParameter("@JobTypeId", @ref, true);
                        _database.AddParameter("@CustomerId", ref2, true);
                        dt = _database.ExecuteSP_DataTable("CostCentre_Get_ByJobTypeIdAndCustomerId");
                        break;
                    }

                default:
                    {
                        _database.AddParameter("@Id", @ref, true);
                        dt = _database.ExecuteSP_DataTable("CostCentre_Get_ById");
                        break;
                    }
            }

            if (dt is object && dt.Rows.Count > 0)
            {
                var costCentres = ObjectMap.DataTableToList<CostCentre>(dt);
                return costCentres;
            }
            else
            {
                return null;
            }
        }

        public CostCentre Insert(CostCentre ccm)
        {
            AddParameters(ccm);
            ccm.Id = Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("CostCentre_Insert"));
            return ccm;
        }

        private void AddParameters(CostCentre ccm)
        {
            _database.ClearParameter();
            _database.AddParameter("@CostCentre", ccm.CostCentreId);
            _database.AddParameter("@JobTypeId", ccm.JobTypeId);
            _database.AddParameter("@LinkId", ccm.LinkId);
            _database.AddParameter("@LinkTypeId", ccm.LinkTypeId);
            var spendLimit = ccm.JobSpendLimit > 0 ? (object)ccm.JobSpendLimit : DBNull.Value;
            _database.AddParameter("@JobSpendLimit", spendLimit, true);
        }

        public CostCentre Update(CostCentre ccm)
        {
            AddParameters(ccm);
            _database.AddParameter("@Id", ccm.Id);
            ccm.Id = Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("CostCentre_Update"));
            return ccm;
        }

        public bool Delete(int Id)
        {
            _database.ClearParameter();
            _database.AddParameter("@Id", Id);
            return Conversions.ToBoolean(_database.ExecuteSP_ReturnRowsAffected("CostCentre_Delete") == 1);
        }
    }
}